Add first draft of the documentation for the breakpoint abstract classes

titouanmathis opened the pull request:
This PR introduces mechanisms to easily work with breakpoints when creating components:

- The `BreakpointObserver` abstract class allows a component to self mount/destroy by breakpoint
- The `BreakpointManager` abstract class allows a component to switch between different component classes by breakpoint

perruche commented on docs/abstracts/BreakpointManager.md:
```diff
@@ -0,0 +1,40 @@
+import MenuMobile from './MenuMobile';
+import MenuDesktop from './MenuDesktop';
+
+export default class Menu extends BreakpointManager {
```
I find it strange to not extend `Base` anymore

lusimeon replied:
I think the inheritance isn't the right pattern. Is it possible to use composition (something similar to [React Hooks](https://fr.reactjs.org/docs/hooks-intro.html) for example) ?
```js
import { BreakpointManager } from '@studiometa/js-toolkit/abstracts';
class Menu { … }
export default BreakpointManager(Menu);
```

titouanmathis replied:
> I find it strange to not extend Base anymore

I would like to avoid having the `Base` class being a Swiss army knife as it would become a nightmare to maintain and test. I find it simpler to extend its functionality with small classes. The `BreakpointManager` and `BreakpointObserver` classes are not compatible, implementing their behavior in one single class would add too much complexity to the code base.

I actually find that different classes can help browse and understand the code quickly, for example: 

- if the component is extending `BreakpointObserver`, I can set the `activeBreakpoints` property
- if the component is extending `BreakpointManager`, I know that I will have to look for different classes based on the breakpoint I am working on

> I think the inheritance isn't the right pattern. Is it possible to use composition (something similar to React Hooks for example) ?

Interesting point, but as far as I could think about it, I don't see any way for now to not use inheritance. We can discuss this in more details if you want 🙂 

I added the two classes to the PR ([`BreakpointManager`](https://github.com/studiometa/js-toolkit/blob/feature%2Fbreakpoint-abstracts-classes/src/abstracts/BreakpointManager.js) and [`BreakpointObserver`](https://github.com/studiometa/js-toolkit/blob/feature%2Fbreakpoint-abstracts-classes/src/abstracts/BreakpointObserver.js)) and added examples to [the demo](https://deploy-preview-27--studiometa-js-toolkit-demo.netlify.app/).

perruche commented on docs/abstracts/BreakpointManager.md:
```diff
@@ -0,0 +1,40 @@
+      name: 'Menu',
+      breakpoints: [
+        ['xxs xs s', MenuMobile],
+        ['m l xl xxl', MenuDesktop],
```
This is real nice

perruche commented on docs/abstracts/BreakpointObserver.md:
```diff
@@ -0,0 +1,50 @@
+  get config() {
+    return {
+      name: 'MobileComponent',
+      inactiveBreakpoints: 'm l xl xxl',
```
This will probably need to throw a conprehensive error if a breakpoint is both in the active and inactive string

titouanmathis replied:
An error is thrown when you configure both the `activeBreakpoints` and `inactiveBreakpoints` at the same time. The message says that the two options are not compatible.

perruche reviewed (commented):
Real nice addition. I'm just wondering about the extend class, is it too much to have it as a feature in the `Base` class ?

titouanmathis commented:
I did some test and it would indeed be practical to have the `BreakpointObserver` class merged into the `Base` to be able to easily enable/disable a component based on the breakpoint without having to create a new class. 

For example, I tried to disable the `Modal` component from the toolkit on small devices. I can not do this without creating a new component based on `BreakpointObserver` to wrap the `Modal` component. 

I will try to find a nice implementation.

perruche commented:
> I did some test and it would indeed be practical to have the `BreakpointObserver` class merged into the `Base` to be able to easily enable/disable a component based on the breakpoint without having to create a new class.
> 
> For example, I tried to disable the `Modal` component from the toolkit on small devices. I can not do this without creating a new component based on `BreakpointObserver` to wrap the `Modal` component.
> 
> I will try to find a nice implementation.

Too bad interfaces aren't a thing in javascript cause something like
`class Foo extends Base implements BreakpointObserver {}` would be awesome

titouanmathis commented:
I updated the proposal and converted the 2 abstract classes into decorators. See the following links for more informations:

- Updated documentation: 
  - [`withBreakpointObserver`](https://deploy-preview-27--studiometa-js-toolkit.netlify.app/decorators/withBreakpointObserver.html)
  - [`withBreakpointManager`](https://deploy-preview-27--studiometa-js-toolkit.netlify.app/decorators/withBreakpointManager.html)
- Updated demo: 
  - [Demo preview](https://deploy-preview-27--studiometa-js-toolkit-demo.netlify.app/) 
  - Sources: [app.js](https://github.com/studiometa/js-toolkit/blob/feature/breakpoint-abstracts-classes/demo/src/js/app.js#L17), [BreakpointObserverDemo.js](https://github.com/studiometa/js-toolkit/blob/feature/breakpoint-abstracts-classes/demo/src/js/components/BreakpointObserverDemo.js) and [BreakpointManagerDemo/index.js](https://github.com/studiometa/js-toolkit/blob/feature/breakpoint-abstracts-classes/demo/src/js/components/BreakPointManagerDemo/index.js)

perruche commented on docs/decorators/withBreakpointObserver.md:
```diff
@@ -0,0 +1,75 @@
+
+# withBreakpointObserver
+
+Use this decorator to create a that will have the capacity to be mounted or destroyed when the current breakpoint changes.
```
"to create a **Class**"

perruche commented on docs/decorators/withBreakpointObserver.md:
```diff
@@ -0,0 +1,75 @@
+
+### Enable for the given breakpoints
+
+In the following example, the `MobileComponent` class will self mount on devices matching any of the breakpoint defined in the `activeBreakpoint` property, and self destroy on all others:
```
Missing the s in `activeBreakpoints`

perruche commented on docs/decorators/withBreakpointObserver.md:
```diff
@@ -0,0 +1,75 @@
+      name: 'App',
+      components: {
+        Modal,
+        ResponsiveModal: withBreakpointObserver(Modal),
```
That's a great behavior 👍
